# Data Understanding
Objetivo do Notebook: fornecer uma visão geral das bases de dados disponíveis, identificar ou criar variáveis úteis, entender a granularidade dos dados, possíveis junções e realizar uma limpeza inicial que não dependa de aprofundamento na base.
Refinamento dos datasets de entrada em `data/01_raw` para os datasets em `data/02_primary`.

Toda a exploração dos dados como **frequências de variáveis**, **correlações** e **visualizações gráficas** será feita a partir dos datasets em `data/02_primary` no notebook `data_exploration.ipynb`.

### Kedro

In [1]:
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from kedro.io import DataCatalog
    catalog: DataCatalog

In [2]:
%load_ext kedro.ipython

[09/28/25 20:52:45] INFO     Using 'c:\Users\Zeus\Projetos\Projetos DS\Projetos                     __init__.py:270
                             Preditiva\otimizacao-de-ads-para-startup\.venv\Lib\site-packages\kedro                
                             \framework\project\rich_logging.yml' as logging configuration.                        

                    INFO     Registered line magic '%reload_kedro'                                   __init__.py:63

                    INFO     Registered line magic '%load_node'                                      __init__.py:65

                    INFO     Resolved project path as: c:\Users\Zeus\Projetos\Projetos DS\Projetos  __init__.py:180
                             Preditiva\otimizacao-de-ads-para-startup.                                             
                             To set a different path, run '%reload_kedro <project_root>'                           

[09/28/25 20:52:47] INFO     Kedro is sending anonymous usage data with the sole purpose of improving plugin.py:243
                             the product. No personal data or IP addresses are stored on our side. To              
                             opt out, set the `KEDRO_DISABLE_TELEMETRY` or `DO_NOT_TRACK` environment              
                             variables, or create a `.telemetry` file in the current working                       
                             directory with the contents `consent: false`. To hide this message,                   
                             explicitly grant or deny consent. Read more at                                        
                             https://docs.kedro.org/en/stable/configuration/telemetry.html                         

                    INFO     Kedro project Otimizacao de Ads para Startup                           __init__.py:146

                    INFO     Defined global variable 'context', 'session', 'catalog' and            __init__.py:147
                             'pipelines'                                                                           

[09/28/25 20:52:48] INFO     Registered line magic 'run_viz'                                        __init__.py:153

### Importações de Bibliotecas

In [3]:

# Importando as bibliotecas necessárias
import warnings

import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np

In [4]:

# Configurações
warnings.filterwarnings("ignore")
plt.style.use("seaborn-v0_8")
sns.set_palette("husl")
plt.rcParams["figure.figsize"] = (12, 8)
plt.rcParams["font.size"] = 12

# Configurando para exibir todas as colunas
pd.set_option("display.max_columns", None)
pd.set_option("display.width", None)


## Visão inicial das bases disponíveis

Temos 4 abas na planilha disponibilizada e faremos a compreensão de cada aba individualmente. 

Nomeei as bases de acordo com o que inicialmente aparentam conter de informações, sendo elas: Metadados, Campanha, Categorias e Categoria Detalhada.

#### Metadados

In [5]:
df_campanha_metadados = catalog.load("metadados")
df_campanha_metadados

[09/28/25 20:52:49] INFO     Loading data from metadados (ExcelDataset)...                     data_catalog.py:1046

Variável  \
0               id_do_anuncio   
1                    campanha   
2                faixa_etária   
3                        sexo   
4      categoria_de_interesse   
5           qte_de_impressões   
6               qte_de_clicks   
7  valor_investido_no_anúncio   
8   qte_de_Vendas_após_Clique   

                                           Descrição  
0  Identificação única do anúncio mostrado na tim...  
1  Campanha realizada. Uma campanha é uma combina...  
2  Faixa de idade do lead em que o anúncio foi mo...  
3        Sexo do lead em que o anúncio foi mostrado.  
4  Identificação da categoria de interesse do lea...  
5  Quantidade de visualizações que este anúncio o...  
6     Quantidade de cliques que este anúncio obteve.  
7  Uma das formas de custo do Facebook Ads é o cu...  
8  Quantidade de vendas de cursos obtidos após o ...

In [6]:
# Aumentar o limite de largura da coluna para melhor visualização
pd.set_option("display.max_colwidth", None)
df_campanha_metadados

Variável  \
0               id_do_anuncio   
1                    campanha   
2                faixa_etária   
3                        sexo   
4      categoria_de_interesse   
5           qte_de_impressões   
6               qte_de_clicks   
7  valor_investido_no_anúncio   
8   qte_de_Vendas_após_Clique   

                                                                                                                                              Descrição
0                                                                                      Identificação única do anúncio mostrado na timeline do Facebook.
1                                                       Campanha realizada. Uma campanha é uma combinação de públicos alvo para se mostrar os anúncios.
2                                                                                                 Faixa de idade do lead em que o anúncio foi mostrado.
3                                                                                                           Sexo do lead em que o anúncio foi mostrado.
4                                                                        Identificação da categoria de interesse do lead em que o anúncio foi mostrado.
5                                                                                                  Quantidade de visualizações que este anúncio obteve.
6                                                                                                        Quantidade de cliques que este anúncio obteve.
7  Uma das formas de custo do Facebook Ads é o custo por clique. Esta variável mede o valor que o Facebook cobrou pela quantidade de cliques recebidos.
8                                                                                      Quantidade de vendas de cursos obtidos após o clique no anúncio.

**Metadados** então tem apenas um detalhamento do significado das colunas disponíveis na aba com os dados de campanha.

Um ponto importante aqui é a coluna de valor_investido_no_anúncio que a descrição informa que o custo é por cliques e não pode visualizações.

In [7]:
# Voltando a visualização padrão
pd.set_option("display.max_colwidth", 50)

#### Campanha

In [8]:
df_campanha = catalog.load("campanhas")
df_campanha.head()

                    INFO     Loading data from campanhas (ExcelDataset)...                     data_catalog.py:1046

id_do_anuncio    campanha faixa_etária sexo  categoria_de_interesse  \
0         708746  Campanha A        30-34    M                      15   
1         708749  Campanha A        30-34    M                      16   
2         708771  Campanha A        30-34    M                      20   
3         708815  Campanha A        30-34    M                      28   
4         708818  Campanha A        30-34    M                      28   

   qte_de_impressões  qte_de_clicks  valor_investido_no_anúncio  \
0               7350              1                        1.43   
1              17861              2                        1.82   
2                693              0                        0.00   
3               4259              1                        1.25   
4               4133              1                        1.29   

   Qte_de_Vendas_após_Clique  
0                          3  
1                          2  
2                          0  
3                          1  
4                          2

In [9]:
df_campanha = df_campanha.rename(columns={
    "qte_de_impressões": "impressoes",
    "qte_de_clicks": "cliques",
    "sexo": "genero",
    "valor_investido_no_anúncio": "custo",
    "Qte_de_Vendas_após_Clique": "vendas"
})

In [10]:
from unidecode import unidecode

df_campanha.columns = [unidecode(col) for col in df_campanha.columns]

In [11]:
df_campanha.head()

id_do_anuncio    campanha faixa_etaria genero  categoria_de_interesse  \
0         708746  Campanha A        30-34      M                      15   
1         708749  Campanha A        30-34      M                      16   
2         708771  Campanha A        30-34      M                      20   
3         708815  Campanha A        30-34      M                      28   
4         708818  Campanha A        30-34      M                      28   

   impressoes  cliques  custo  vendas  
0        7350        1   1.43       3  
1       17861        2   1.82       2  
2         693        0   0.00       0  
3        4259        1   1.25       1  
4        4133        1   1.29       2

Como o foco aqui é apenas entender o que há nas bases, não vou me aprofundar muito mais. Apenas ajustamos alguns nomes de colunas e tiramos acentos pra ficarem de mais fácil acesso.

#### Categorias

In [12]:
df_categorias = catalog.load("categorias")
df_categorias.head()

                    INFO     Loading data from categorias (ExcelDataset)...                    data_catalog.py:1046

categoria_de_interesse  descrição_da_categoria
0                       2  Business and Industry 
1                       7                     NaN
2                      10                     NaN
3                      15                     NaN
4                      16          Entertainment

Estranho esses valores nulos..

In [13]:
df_categorias.columns = [unidecode(col) for col in df_categorias.columns]
df_categorias

categoria_de_interesse     descricao_da_categoria
0                        2     Business and Industry 
1                        7                        NaN
2                       10                        NaN
3                       15                        NaN
4                       16             Entertainment 
5                       18                        NaN
6                       19                        NaN
7                       20                        NaN
8                       21  Family and relationships 
9                       22                        NaN
10                      23                        NaN
11                      24                        NaN
12                      25      Fitness and wellness 
13                      26                        NaN
14                      27                        NaN
15                      28                        NaN
16                      29            Food and drink 
17                      30                        NaN
18                      31                        NaN
19                      32                        NaN
20                      36    Hobbies and activities 
21                      63                        NaN
22                      64                        NaN
23                      65                        NaN
24                      66      Shopping and fashion 
25                     100                        NaN
26                     101                        NaN
27                     102                        NaN
28                     103       Sports and outdoors 
29                     104                        NaN
30                     105                        NaN
31                     106                        NaN
32                     107                Technology 
33                     108                        NaN
34                     109                        NaN
35                     110                        NaN
36                     111                        NaN
37                     112                        NaN
38                     113                        NaN
39                     114                        NaN

Provável que estivessem usando celulas mescladas no excel, tendo texto apenas na primeira linha da mescla. Um ffill resolve isso.

#### Categoria Detalhada

In [14]:
df_categoria_detalhada = catalog.load("categoria_detalhada")
df_categoria_detalhada.head(20)

[09/28/25 20:52:50] INFO     Loading data from categoria_detalhada (ExcelDataset)...           data_catalog.py:1046

categoria_detalhada
0                              Business and Industry 
1                                                 NaN
2   Advertising Agriculture Architecture Aviation ...
3                                                 NaN
4   Investment banking Online banking Retail banking 
5                                                 NaN
6                       Business Construction Design 
7                                                 NaN
8      Fashion design Graphic design Interior design 
9                                                 NaN
10  Economics Engineering Entrepreneurship Health ...
11                                                NaN
12                                             Online
13                                 Digital marketing 
14                                                NaN
15                                Display advertising
16                                    Email marketing
17                                 Online advertising
18           Search engine optimization Social media 
19                                                NaN

- *Comportamento estranho das categorias*

A base de **Categoria** parece ser mais abrangente e a **Planilha 8** mais detalhada, por isso nomeei como **Categoria Detalhada**. Porém ambas tem diversos registros nulos e nas categoria detalhada não há uma referência de número da categoria. (os números exibidos na esquerda da exibição anterior são apenas o index gerado automaticamente pelo pandas).

Olhando pela planilha, na aba que chamamos de **Categoria** parece que Business and Industry se aplicaria às próximas 3 linhas também, que poderia ser usado um ffill pra preencher, mas assim teriamos diversos números de categorias de interesse com descrição de categoria repetida. Enquanto se olharmos a planilha 8, dá pra ver que dentro de business industry aparentam ter 17 outras quebras, ou um tree map, onde as linhas sequenciadas representam subcategorias, mas que não se encaixaria nas 3 descrições nulas da aba de categorias.

### Como cruzar os códigos de categorias com suas descrições?

In [15]:
df_campanha["categoria_de_interesse"] = df_campanha["categoria_de_interesse"].astype("category")
df_campanha[["categoria_de_interesse"]].describe()

categoria_de_interesse
count                     1143
unique                      40
top                         16
freq                       140

In [16]:
df_categoria_detalhada.describe()

categoria_detalhada
count                      148
unique                     148
top     Business and Industry 
freq                         1

In [17]:
df_categorias["descricao_da_categoria"].describe()


count                          9
unique                         9
top       Business and Industry 
freq                           1
Name: descricao_da_categoria, dtype: object

In [18]:
df_categorias["descricao_da_categoria"].ffill().describe()


count              40
unique              9
top       Technology 
freq                8
Name: descricao_da_categoria, dtype: object

Dentre as categoiras presentes na base **Campanha**, são 40 valores únicos. Nas **Categorias Detalhadas** existem 148 descrições e em **Categorias** são 9.

Porém realizando o preenchimento de ffill, como eu havia mencionado anteriormente, ficamos com 40 códigos de categorias, igual ao número presente nas campanhas. Por mais que não serão descrições únicas.

- Por isso, usarei a base **Categorias** aplicando ffill para juntar a descrição das categorias à base de detalhes das campanhas.

In [19]:
df_categorias["descricao_da_categoria"] = df_categorias["descricao_da_categoria"].ffill()
df_categorias

categoria_de_interesse     descricao_da_categoria
0                        2     Business and Industry 
1                        7     Business and Industry 
2                       10     Business and Industry 
3                       15     Business and Industry 
4                       16             Entertainment 
5                       18             Entertainment 
6                       19             Entertainment 
7                       20             Entertainment 
8                       21  Family and relationships 
9                       22  Family and relationships 
10                      23  Family and relationships 
11                      24  Family and relationships 
12                      25      Fitness and wellness 
13                      26      Fitness and wellness 
14                      27      Fitness and wellness 
15                      28      Fitness and wellness 
16                      29            Food and drink 
17                      30            Food and drink 
18                      31            Food and drink 
19                      32            Food and drink 
20                      36    Hobbies and activities 
21                      63    Hobbies and activities 
22                      64    Hobbies and activities 
23                      65    Hobbies and activities 
24                      66      Shopping and fashion 
25                     100      Shopping and fashion 
26                     101      Shopping and fashion 
27                     102      Shopping and fashion 
28                     103       Sports and outdoors 
29                     104       Sports and outdoors 
30                     105       Sports and outdoors 
31                     106       Sports and outdoors 
32                     107                Technology 
33                     108                Technology 
34                     109                Technology 
35                     110                Technology 
36                     111                Technology 
37                     112                Technology 
38                     113                Technology 
39                     114                Technology

In [20]:
df_campanha_merged = (
    df_campanha.merge(
        df_categorias,
        how="left",
        on="categoria_de_interesse"
    )
)

In [21]:
df_campanha_merged.head()

id_do_anuncio    campanha faixa_etaria genero  categoria_de_interesse  \
0         708746  Campanha A        30-34      M                      15   
1         708749  Campanha A        30-34      M                      16   
2         708771  Campanha A        30-34      M                      20   
3         708815  Campanha A        30-34      M                      28   
4         708818  Campanha A        30-34      M                      28   

   impressoes  cliques  custo  vendas  descricao_da_categoria  
0        7350        1   1.43       3  Business and Industry   
1       17861        2   1.82       2          Entertainment   
2         693        0   0.00       0          Entertainment   
3        4259        1   1.25       1   Fitness and wellness   
4        4133        1   1.29       2   Fitness and wellness

Com isso, todos os dados que temos disponíveis estão disponíveis na base **df_campanha_merged**, vamos fazer uma análise superficial de como esta base está.

## df_campanha_merged

### Análise Descritiva Superficial

In [22]:
df_campanha_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1143 entries, 0 to 1142
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id_do_anuncio           1143 non-null   int64  
 1   campanha                1143 non-null   object 
 2   faixa_etaria            1143 non-null   object 
 3   genero                  1143 non-null   object 
 4   categoria_de_interesse  1143 non-null   int64  
 5   impressoes              1143 non-null   int64  
 6   cliques                 1143 non-null   int64  
 7   custo                   1143 non-null   float64
 8   vendas                  1143 non-null   int64  
 9   descricao_da_categoria  1143 non-null   object 
dtypes: float64(1), int64(5), object(4)
memory usage: 89.4+ KB


In [23]:
df_campanha_merged = df_campanha_merged.astype({
    "campanha": "category",
    "faixa_etaria": "category",
    "genero": "category",
    "categoria_de_interesse": "category",
    "descricao_da_categoria": "category"
})
df_campanha_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1143 entries, 0 to 1142
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype   
---  ------                  --------------  -----   
 0   id_do_anuncio           1143 non-null   int64   
 1   campanha                1143 non-null   category
 2   faixa_etaria            1143 non-null   category
 3   genero                  1143 non-null   category
 4   categoria_de_interesse  1143 non-null   category
 5   impressoes              1143 non-null   int64   
 6   cliques                 1143 non-null   int64   
 7   custo                   1143 non-null   float64 
 8   vendas                  1143 non-null   int64   
 9   descricao_da_categoria  1143 non-null   category
dtypes: category(5), float64(1), int64(4)
memory usage: 52.5 KB


In [24]:
# Visualização personalizada de pacote pessoa
from utils import value_counts, full_describe

In [25]:
full_describe(df_campanha_merged)

### DESCRIÇÃO COMPLETA

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1143 entries, 0 to 1142
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype   
---  ------                  --------------  -----   
 0   id_do_anuncio           1143 non-null   int64   
 1   campanha                1143 non-null   category
 2   faixa_etaria            1143 non-null   category
 3   genero                  1143 non-null   category
 4   categoria_de_interesse  1143 non-null   category
 5   impressoes              1143 non-null   int64   
 6   cliques                 1143 non-null   int64   
 7   custo                   1143 non-null   float64 
 8   vendas                  1143 non-null   int64   
 9   descricao_da_categoria  1143 non-null   category
dtypes: category(5), float64(1), int64(4)
memory usage: 52.5 KB


<pandas.io.formats.style.Styler object at 0x00000236AC052DD0>

<pandas.io.formats.style.Styler object at 0x00000236AC30ABD0>

<pandas.io.formats.style.Styler object at 0x00000236AC3B7B10>

- **Nulos**: Nada nulo.
- **Cardinalidade**: 3 campanhas, 4 faixas etárias, 40 categorias de interesse

Custo e Impressões tem um alto Coeficiente de Variação (Desvio padrão / Média), indicando alta dispersão dos valores.

Pela visualização inicial dá pra perceber que os índices 3 e 4 são o mesmo perfil e mesma campanha mas em um id de anúncio diferente. Vou considerar que estes anúncios tiveram criativos diferentes exibidos e por isso tiveram ids diferentes pro mesmo perfil na mesma campanha.
Porém nossa base não tem informações sobre o designe dos criativos pra que possamos avaliar qual seria o melhor criativo ou encorpar mais nossa análise. Provável que sigamos com a remoção dessa coluna.



In [26]:
value_counts(df_campanha_merged)

### VALORES ÚNICOS
__apenas top 10 valores__

<pandas.io.formats.style.Styler object at 0x00000236AC3B7550>

<pandas.io.formats.style.Styler object at 0x00000236AC2E9C50>

<pandas.io.formats.style.Styler object at 0x00000236ABF21DD0>

<pandas.io.formats.style.Styler object at 0x00000236AC43C810>

<pandas.io.formats.style.Styler object at 0x00000236AC480190>

### Desempenho de diferentes anuncios do mesmo perfil.

Como queremos identificar os melhores perfis pra uma **nova campanha**, acredito que seja melhor remover a coluna __campanha__ e __id_do_anuncio__ pra agrupar tudo por perfil (faixa_etária, gênero e descrição_da_categoria).

Mas antes, quero validar se o desempenho dos perfis são parecidos em diferentes campanhas.

In [27]:
df_campanha_merged["perfil"] = (
    df_campanha_merged["faixa_etaria"].astype(str) + " | " +
    df_campanha_merged["genero"].astype(str) + " | " +
    df_campanha_merged["categoria_de_interesse"].astype(str)
).astype("category")

Pra avaliar os perfis, vou criar alguns KPIs comumente usados em marketing

In [28]:
# Valor de venda do curso definido na apresentação do problema (disponível no Readme).
VALOR_VENDA = 85.00

# Calculando os KPIs
df_campanha_merged["faturamento"] = df_campanha_merged["vendas"] * VALOR_VENDA   # Receita total gerada pelas vendas
df_campanha_merged["lucro"] = df_campanha_merged["faturamento"] - df_campanha_merged["custo"]   # Lucro líquido = receita - custo do anúncio
df_campanha_merged["ctr"] = (df_campanha_merged["cliques"] / df_campanha_merged["impressoes"])   # Click Through Rate (%) = % de impressões que geraram clique
df_campanha_merged["tc"] = (df_campanha_merged["vendas"] / df_campanha_merged["cliques"]) # Taxa de conversão (%) = % de cliques que geraram venda
df_campanha_merged["cpc"] = df_campanha_merged["custo"] / df_campanha_merged["cliques"]   # Custo por clique médio
df_campanha_merged["cc"] = df_campanha_merged["custo"] / df_campanha_merged["vendas"]   # Custo por conversão (venda)
df_campanha_merged["roi"] = (df_campanha_merged["faturamento"] - df_campanha_merged["custo"]) / df_campanha_merged["custo"]   # Retorno sobre investimento (%) = lucro / custo * 100
df_campanha_merged["conv_i"] = df_campanha_merged["vendas"] / df_campanha_merged["impressoes"]  # Taxa de conversão por impressão

In [29]:
df_campanha_merged.sort_values(by="perfil").head(20)

id_do_anuncio    campanha faixa_etaria genero categoria_de_interesse  \
468          951033  Campanha B        30-34      F                     10   
825         1121742  Campanha C        30-34      F                     10   
173          747791  Campanha B        30-34      F                     10   
469          951035  Campanha B        30-34      F                     10   
234          776579  Campanha B        30-34      F                     10   
77           734785  Campanha B        30-34      F                     10   
172          747790  Campanha B        30-34      F                     10   
824         1121741  Campanha C        30-34      F                     10   
1091        1314357  Campanha C        30-34      F                    101   
1092        1314358  Campanha C        30-34      F                    102   
1093        1314359  Campanha C        30-34      F                    103   
1094        1314360  Campanha C        30-34      F                    104   
1095        1314361  Campanha C        30-34      F                    105   
1096        1314362  Campanha C        30-34      F                    106   
1097        1314363  Campanha C        30-34      F                    107   
1098        1314364  Campanha C        30-34      F                    108   
1099        1314365  Campanha C        30-34      F                    109   
1100        1314366  Campanha C        30-34      F                    110   
1101        1314367  Campanha C        30-34      F                    111   
1102        1314368  Campanha C        30-34      F                    112   

      impressoes  cliques       custo  vendas  descricao_da_categoria  \
468         5517        1    1.230000       1  Business and Industry    
825       213016       30   44.220000      13  Business and Industry    
173        31393        8   10.960001       2  Business and Industry    
469         1539        0    0.000000       0  Business and Industry    
234        26910        5    7.230000       1  Business and Industry    
77          5576        1    1.530000       2  Business and Industry    
172         2077        0    0.000000       0  Business and Industry    
824       318042       46   64.410000       9  Business and Industry    
1091      524306       81  113.680003      15   Shopping and fashion    
1092      104496        9   11.430000       7   Shopping and fashion    
1093      452519       68   99.520002       7    Sports and outdoors    
1094      442919       76  110.780002      23    Sports and outdoors    
1095      596831       86  120.879999      11    Sports and outdoors    
1096      173912       26   35.540000       5    Sports and outdoors    
1097      780967       86  119.640002      26             Technology    
1098      132124        8   11.190000       3             Technology    
1099      623137      100  138.920001      12             Technology    
1100       99020       10   14.480000       4             Technology    
1101      665817      117  163.800000      33             Technology    
1102      699232       80  111.989999      15             Technology    

               perfil  faturamento        lucro       ctr        tc       cpc  \
468    30-34 | F | 10         85.0    83.770000  0.000181  1.000000  1.230000   
825    30-34 | F | 10       1105.0  1060.780000  0.000141  0.433333  1.474000   
173    30-34 | F | 10        170.0   159.039999  0.000255  0.250000  1.370000   
469    30-34 | F | 10          0.0     0.000000  0.000000       NaN       NaN   
234    30-34 | F | 10         85.0    77.770000  0.000186  0.200000  1.446000   
77     30-34 | F | 10        170.0   168.470000  0.000179  2.000000  1.530000   
172    30-34 | F | 10          0.0     0.000000  0.000000       NaN       NaN   
824    30-34 | F | 10        765.0   700.590000  0.000145  0.195652  1.400217   
1091  30-34 | F | 101       1275.0  1161.319997  0.000154  0.185185  1.403457   
1092  30-34 | F | 102  

Podemos identificar acima que os desempenhos de cada design num mesmo perfil tiveram diferenças relevantes. Como não temos uma base relacionando quais designs foram usados em quais id_de_anuncios ou qualquer que seja a diferença entre os anúncios da mesma campanha e perfil, não temos como inferir muita coisa.

**Portanto, seguiremos com apenas as informações de perfil e o desempenho de cada um deles.**

In [30]:
df_perfil = (
    df_campanha_merged
        [["perfil",
          "faixa_etaria",
          "genero",
          "categoria_de_interesse",
          "descricao_da_categoria",
          "impressoes",
          "cliques",
          "custo",
          "vendas"]]
    .groupby(
        ["faixa_etaria",
         "genero",
         "categoria_de_interesse",
         "descricao_da_categoria",
         "perfil"],
         observed=True
    )
    .agg("sum")
    .reset_index()
)

KPI's recriados com novos dados agrupados por perfil.

In [31]:
df_perfil["faturamento"] = df_perfil["vendas"] * VALOR_VENDA   # Receita total gerada pelas vendas
df_perfil["lucro"] = df_perfil["faturamento"] - df_perfil["custo"]   # Lucro líquido = receita - custo do anúncio
df_perfil["ctr"] = (df_perfil["cliques"] / df_perfil["impressoes"])   # Click Through Rate (%) = % de impressões que geraram clique
df_perfil["tc"] = (df_perfil["vendas"] / df_perfil["cliques"]) # Taxa de conversão (%) = % de cliques que geraram venda
df_perfil["cpc"] = df_perfil["custo"] / df_perfil["cliques"]   # Custo por clique médio
df_perfil["cc"] = df_perfil["custo"] / df_perfil["vendas"]   # Custo por conversão (venda)
df_perfil["roi"] = (df_perfil["faturamento"] - df_perfil["custo"]) / df_perfil["custo"]   # Retorno sobre investimento (%) = lucro / custo * 100
df_perfil["conversao"] = df_perfil["vendas"] / df_perfil["impressoes"]  # Taxa de conversão por impressão

## df_perfil

### Análise Descritiva Superficial

In [32]:
full_describe(df_perfil)

### DESCRIÇÃO COMPLETA

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284 entries, 0 to 283
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype   
---  ------                  --------------  -----   
 0   faixa_etaria            284 non-null    category
 1   genero                  284 non-null    category
 2   categoria_de_interesse  284 non-null    category
 3   descricao_da_categoria  284 non-null    category
 4   perfil                  284 non-null    category
 5   impressoes              284 non-null    int64   
 6   cliques                 284 non-null    int64   
 7   custo                   284 non-null    float64 
 8   vendas                  284 non-null    int64   
 9   faturamento             284 non-null    float64 
 10  lucro                   284 non-null    float64 
 11  ctr                     284 non-null    float64 
 12  tc                      281 non-null    float64 
 13  cpc                     281 non-null    float64 
 14  cc                      28

<pandas.io.formats.style.Styler object at 0x00000236AC3EEAD0>

<pandas.io.formats.style.Styler object at 0x00000236AC3F2710>

<pandas.io.formats.style.Styler object at 0x00000236AC3EB4D0>

Agora temos 284 perfis únicos sendo eles a combinação entre faixa etária, gênero e categoria de interesse.
Desta vez estamos avaliando também os detalhes das métricas de KPI's que geramos.

Temos que nos atentar aos valores infinitos gerados em cc por conta da divisão por zero. Vamos substituir por nulo

In [35]:
df_perfil["cc"] = df_perfil["cc"].replace([np.inf, -np.inf], np.nan)

Vou rodar o full_describe de novo só pra visualizar métricas do KPI cc que antes estavam comprometidos por conta dos valores infinitos 

In [36]:
full_describe(df_perfil[["cc"]])

### DESCRIÇÃO COMPLETA

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284 entries, 0 to 283
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   cc      274 non-null    float64
dtypes: float64(1)
memory usage: 2.3 KB


<pandas.io.formats.style.Styler object at 0x00000236AC456810>

<pandas.io.formats.style.Styler object at 0x00000236AC455A50>

In [37]:
value_counts(df_perfil.drop(columns=["perfil"]))

### VALORES ÚNICOS
__apenas top 10 valores__

<pandas.io.formats.style.Styler object at 0x00000236AC43B550>

<pandas.io.formats.style.Styler object at 0x00000236AC2FDC90>

<pandas.io.formats.style.Styler object at 0x00000236AC4568D0>

<pandas.io.formats.style.Styler object at 0x00000236AC3E8410>

Nenhuma das categorias tem desbalanços problemáticos, apenas a descrição de categoria Technology que é aplicada á mais categorias de interesse, mas sem problemas.

In [37]:
df_perfil.head()

faixa_etaria genero categoria_de_interesse  descricao_da_categoria  \
0        30-34      F                      2  Business and Industry    
1        30-34      F                      7  Business and Industry    
2        30-34      F                     10  Business and Industry    
3        30-34      F                     15  Business and Industry    
4        30-34      F                     16          Entertainment    

           perfil  impressoes  cliques       custo  vendas  faturamento  \
0   30-34 | F | 2      252951       45   67.200000      11        935.0   
1   30-34 | F | 7      317046       58   84.860000      12       1020.0   
2  30-34 | F | 10      604070       91  129.580000      28       2380.0   
3  30-34 | F | 15     1669161      262  391.600002      46       3910.0   
4  30-34 | F | 16     2693782      355  529.709996      62       5270.0   

         lucro       ctr        tc       cpc        cc        roi  conversao  
0   867.800000  0.000178  0.244444  1.493333  6.109091  12.913690   0.000043  
1   935.140000  0.000183  0.206897  1.463103  7.071667  11.019797   0.000038  
2  2250.420000  0.000151  0.307692  1.423956  4.627857  17.367032   0.000046  
3  3518.399998  0.000157  0.175573  1.494656  8.513044   8.984678   0.000028  
4  4740.290004  0.000132  0.174648  1.492141  8.543710   8.948840   0.000023

**df_perfis** será a base que usaremos na próxima etapa de análise exploratória mais detalhada. Com isso salvaremos ela em `data/02_primary`

# Resumo final

Etapas seguidas neste documento:

- Compreensão de bases disponíveis
- Decisão de como juntá-las
- Remoção de variáveis para redefinição de granularidade, deixando a granularidade por perfil
  - um perfil foi definido como a junção de Faixa etária, Gênero e categoria de Interesse
- Criação de variáveis de KPI's comuns em marketing digital